### PR TITLE
Implement TS Laser Fences

### LINKED ISSUE #10790: Laser fences
WIP at https://github.com/teees/openra/tree/laser-fence

### PR DESCRIPTION
**Pull request by pchote**
* Closes #10790.
* Makes our PBOG line-build rendering more closely match TS.
* Fixes the powerdown indicator being rendered too low on all actors.
* `WithMakeAnimation` conditions gives us the plumbing for future PRs to kill all the `INotifyBuildComplete` hacks.

### CODE REVIEW

**Comment by pchote** [comment]
The laser fence segments look weird in the map editor, but that is a problem with the editor / actor previews (which don't support conditions) rather than this PR.  Map-placed laser fences are expected to be set up with the correct `ActorInit`s.  The following snippet can be pasted into the TS shellmap to demonstrate this:

```yaml
	Actor445: napost
		Owner: Creeps
		Location: 57,0
	Actor447: napost
		Owner: Creeps
		Location: 57,5
	Actor442: nafnce
		Owner: Creeps
		Location: 57,1
		LineBuildDirection: Y
		LineBuildParent: Actor445, Actor447
	Actor443: nafnce
		Owner: Creeps
		Location: 57,2
		LineBuildDirection: Y
		LineBuildParent: Actor445, Actor447
	Actor444: nafnce
		Owner: Creeps
		Location: 57,3
		LineBuildDirection: Y
		LineBuildParent: Actor445, Actor447
	Actor446: nafnce
		Owner: Creeps
		Location: 57,4
		LineBuildDirection: Y
		LineBuildParent: Actor445, Actor447
	Actor448: gapowr
		Owner: Creeps
		Location: 51,49
```

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnLineBuildDirection.cs
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnLineBuildDirection.cs]: "build" is missing its 'b'.

**Comment by pchote** [comment]
Fixed and rebased.

**Comment by pchote** [comment]
Simplified the `SegmentsRequireNode` implementation by reordering the commits and using later bits earlier.

**Comment by ltem** [comment]
Tiles,  with a laser fence segment, will allow only one infantry unit placed on them. It seems that this causes path finding issues. If a large group is sent over a deactivated laser fence, some infantry units will stop in front of the fence.

This group of infantry was sent to the other side, some of them stopped.

image
image

**Comment by pchote** [comment]
That sounds like a bug in the generic temporary blocker logic.  I expect that gates will have the same issue?

**Comment by ltem** [comment]
Kind of. If the gate is open, it happens rarely. If the gate needs to open itself, it happens often. Sometimes the units stop in front of the gate, sometimes on the gate-tiles. It isn't very consistent :/

**Comment by pchote** [comment]
That pretty much confirms that the bug isn't in the new laser fence logic, so fixing it will be a job for another PR.

**Comment by pchote** [comment]
This PR is blocking #13051, which affects all mods, so it would be good to get some movement here.

File: OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs
Comment (GraionDilach) [inline OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs]: Why is 2 skipped?

Comment (pchote) [inline OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs]: There used to be another flag there.  Fixed.

**Comment by pchote** [comment]
Rebased, fixed the sidebar icon order (after #13053), and fixed an issue with the building placement palette.

File: OpenRA.Mods.Common/Traits/Render/WithMakeAnimation.cs
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Render/WithMakeAnimation.cs]: `conditionManager != null` should be true if `token != ConditionManager.InvalidConditionToken` is true.
However, that is too minor of a detail to hold this PR off longer.

File: OpenRA.Mods.Cnc/Traits/EnergyWall.cs
Comment (atlimit8) [inline OpenRA.Mods.Cnc/Traits/EnergyWall.cs]: Just noting for rebasing that `ConditionExpression` should be `BooleanExpression` here.

File: OpenRA.Mods.Cnc/Traits/EnergyWall.cs
Comment (atlimit8) [inline OpenRA.Mods.Cnc/Traits/EnergyWall.cs]: Just noting for rebasing that ` > 0` is to be removed since `BooleanExpression.Evaluate(conditions)` returns a bool.

**Comment by pchote** [comment]
Fixed and rebased.

**Review by atlimit8 — APPROVED** [review]
The only issues I see left are:
 * Some actors in a group not crossing which is probably a bug in `ITemporaryBlocker` handling outside of this PR and would benefit from the merge by having another test case.
 * The line builds cannot be teed off of a middle segment or shortened because another post cannot be placed on top, but that is for a some future PR related to #12088.

File: OpenRA.Mods.Common/Traits/Render/WithMakeAnimation.cs
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Render/WithMakeAnimation.cs]: This conflicts with #13041 and shouldn't be necessary anyway (since you're only checking for the trait in INotifyCreated, not the constructor).

File: mods/ts/rules/nod-support.yaml
Comment (reaperrr) [inline mods/ts/rules/nod-support.yaml]: IIRC the posts actually sold for 100 credits in the original, only the segments didn't.

File: mods/ts/rules/nod-support.yaml
Comment (reaperrr) [inline mods/ts/rules/nod-support.yaml]: This is wrong, the post did have a damage transition sound in TS, like the other buildings.

File: mods/ts/rules/nod-support.yaml
Comment (reaperrr) [inline mods/ts/rules/nod-support.yaml]: This is wrong, in the original they did have the same transition sound other buildings had.

**Comment by reaperrr** [comment]
👍 after the 3 minor issues have been adressed.

**Comment by pchote** [comment]
Updated.  I had to change `WithMakeAnimation` significantly to bring it in line with #12919 and what you want to do.  I suspect we are going to need a new PR to remove the `// HACK` before the end of this rework.

**Comment by pchote** [comment]
Rebased after #13074.

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
Looks good to me now, but I'd like @atlimit8 to give the updated `WithMakeAnimation` changes a last look before this is merged.

File: OpenRA.Mods.Common/Traits/Render/WithMakeAnimation.cs
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Render/WithMakeAnimation.cs]: This could be done soon by granting a condition, or, in the future, an expression variable not unlike my `health.Damage == DamageStates.critical` example: `makeState == makeState.Complete`. Other traits would just trigger when the condition is true.

File: OpenRA.Mods.Common/Traits/Render/WithMakeAnimation.cs
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Render/WithMakeAnimation.cs]: Is there a reason that `WithSpriteBody` is cached but not `ConditionManager`.
If `conditionManager = string.IsNullOrEmpty(info.Condition) ? null : self.TraitOrDefault<ConditionManager>();` was in `INotifyCreated.Created(Actor self)`, then this could be reduced to:
```c#
			if (conditionManager != null && token == ConditionManager.InvalidConditionToken)
				token = conditionManager.GrantCondition(self, info.Condition);
```

File: OpenRA.Mods.Common/Traits/Render/WithMakeAnimation.cs
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Render/WithMakeAnimation.cs]: Same here

File: OpenRA.Mods.Common/Traits/Render/WithMakeAnimation.cs
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Render/WithMakeAnimation.cs]: and here

File: OpenRA.Mods.Common/Traits/Render/WithMakeAnimation.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithMakeAnimation.cs]: I'm now caching the condition manager, but i'd prefer not to merge the info.Condition check into the manager query because that breaks the standard pattern that i'd like to keep (it won't work if a trait grants multiple conditions).

**Comment by pchote** [comment]
Updated.

File: OpenRA.Mods.Common/Traits/Render/WithMakeAnimation.cs
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Render/WithMakeAnimation.cs]: OK

**Comment by atlimit8** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/57c098d199d81a5cb5a830d862ff104618989a95...874f85dda7c80ba5d5533f2f073eddc9f5ec45aa)

**Pull request merged**
